server/resp/src/main/java/org/infinispan/server/resp/operation/SetOperation.java

```diff
@@ -0,0 +1,212 @@
+package org.infinispan.server.resp.operation;
+
+import java.nio.charset.StandardCharsets;
+import java.time.Instant;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.TimeUnit;
+
+import org.infinispan.AdvancedCache;
+import org.infinispan.commons.util.concurrent.CompletableFutures;
+import org.infinispan.container.entries.CacheEntry;
+import org.infinispan.server.resp.response.SetResponse;
+
+public class SetOperation {
+   private static final CompletableFuture<Long> NO_EXPIRATION = CompletableFuture.completedFuture(-1L);
+
+   public static CompletionStage<SetResponse> performOperation(AdvancedCache<byte[], byte[]> cache, List<byte[]> arguments) {
+      if (arguments.size() < 2) throw new IllegalArgumentException("SET operation does not have enough arguments");
+
+      SetOperationOptions options = new SetOperationOptions(arguments);
+
+      CompletionStage<Long> expirationTime;
+      // We delay the retrieval for later when operation has `XX`.
+      if (options.isKeepingTtl() && options.typeOfOperation() != SetOperationType.PUT_IF_PRESENT) {
+         expirationTime = cache.getCacheEntryAsync(options.key)
+               .thenApply(e -> e == null ? -1 : extractCurrentTTL(e));
+      } else if (options.isUsingExpiration()) {
+         expirationTime = CompletableFuture.completedFuture(options.expirationMs);
+      } else {
+         expirationTime = NO_EXPIRATION;
+      }
+
+      CompletionStage<byte[]> operation = expirationTime
+            .thenCompose(expiration -> {
+               byte[] key = options.key;
+               byte[] value = options.value;
+
+               switch (options.typeOfOperation()) {
+                  case PUT:
+                     return cache.putAsync(key, value, expiration, TimeUnit.MILLISECONDS);
+                  case PUT_IF_ABSENT:
+                     return cache.putIfAbsentAsync(key, value, expiration, TimeUnit.MILLISECONDS);
+               }
+
+               return cache.getCacheEntryAsync(key)
+                     .thenCompose(entry -> {
+                        if (entry == null || entry.isNull()) {
+                           return CompletableFutures.completedNull();
+                        }
+
+                        long exp = expiration;
+                        if (options.isKeepingTtl()) {
+                           exp = extractCurrentTTL(entry);
+                        }
+
+                        byte[] prev = entry.getValue();
+                        return cache.replaceAsync(key, prev, value, exp, TimeUnit.MILLISECONDS)
+                              .thenApply(b -> b ? prev : null);
+                     });
+            });
+
+      return operation.thenApply(v -> {
+         switch (options.typeOfOperation()) {
+            case PUT_IF_PRESENT:
```

Comment:
I personally don't like this, instead we should just set the correct Consumer based on the type, avoids all these extra unneeded allocations.